test/functional/tests/fault/test_primary_device_error.py

```diff
@@ -0,0 +1,159 @@
+#
+# Copyright(c) 2019 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+import pytest
+
+from test_tools.fio.fio import Fio
+from test_tools.fio.fio_param import ReadWrite, IoEngine, ErrorFilter, VerifyMethod
+from test_tools.device_mapper import ErrorDevice, DmTable
+from core.test_run import TestRun
+from api.cas import casadm
+from api.cas.cache_config import (
+    CacheMode,
+    CacheLineSize,
+    SeqCutOffPolicy,
+    CleaningPolicy,
+    CacheStatus,
+)
+from storage_devices.disk import DiskTypeSet, DiskType, DiskTypeLowerThan
+from storage_devices.device import Device
+from test_utils.os_utils import Udev, sync
+from test_utils.size import Size, Unit
+from test_utils.filesystem.file import File
+
+
+@pytest.mark.parametrize("cache_line_size", CacheLineSize)
+@pytest.mark.parametrize("cache_mode", CacheMode)
+@pytest.mark.parametrize("io_dir", [ReadWrite.randread, ReadWrite.randwrite])
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_core_device_error(io_dir, cache_mode, cache_line_size):
+    """
+        title: Check if CAS behaves correctly when encountering errors on core device
+        description: |
+          Perform I/O on two exported objects created using error and non-error device.
+          Validate CAS that stats counting is consistent with OS reporting.
+          Also, check if normal I/O is uninterrupted and no DC occurs on any of the
+          core devices.
+        pass_criteria:
+          - I/O error count in FIO and in cache statistics match
+          - Positively passed fio verify on both core devices
+    """
+    cache, error_core, error_core_device, good_core, good_core_device = prepare_configuration(
```

Comment:
Isn't error_core_device and good_core_device redundant? You can retrieve those values from error_core and good_core.

Reply:
True that, fixed